src/rgw/rgw_rest_s3.cc

```diff
@@ -3782,6 +3782,173 @@ void RGWPutACLs_ObjStore_S3::send_response()
+       if (iter != attrs.end()) {
+	 lo_etag = iter->second.to_str();
+       }
+      }
+      s->formatter->dump_string("ETag", lo_etag);
+    }
+
+    if (requested_attributes & as_flag(ReqAttributes::Checksum)) {
+      s->formatter->open_object_section("Checksum");
+      auto iter = attrs.find(RGW_ATTR_CKSUM);
+      if (iter != attrs.end()) {
+	try {
+	  rgw::cksum::Cksum cksum;
+	  auto bliter = iter->second.cbegin();
+	  cksum.decode(bliter);
+          if (multipart_parts_count && multipart_parts_count > 0) {
+	    s->formatter->dump_string(cksum.element_name(),
+		fmt::format("{}-{}", cksum.to_armor(), *multipart_parts_count));
+	  } else {
+	    s->formatter->dump_string(cksum.element_name(), cksum.to_armor());
+	  }
+	} catch (buffer::error& err) {
+	  ldpp_dout(this, 0)
+	    << "ERROR: could not decode stored cksum, caught buffer::error" << dendl;
+	}
+      }
+      s->formatter->close_section(); /* Checksum */
+    } /* Checksum */
+
+    if (requested_attributes & as_flag(ReqAttributes::ObjectParts)) {
+      if (multipart_parts_count && multipart_parts_count > 0) {
+
+	/* XXX the following was needed to see a manifest at list_parts()! */
+	op_ret = s->object->load_obj_state(s, s->yield);
+	if (op_ret < 0) {
+	  ldpp_dout_fmt(this, 0,
+			"ERROR: {} load_obj_state() failed ret={}", __func__,
+			op_ret);
+	}
+
+	ldpp_dout_fmt(this, 16,
+		      "{} attr flags={} parts_count={}",
+		      __func__, requested_attributes, *multipart_parts_count);
+
+        s->formatter->open_object_section("ObjectParts");
+
+	bool truncated = false;
+	int marker = 0;
+	int next_marker;
+
+	using namespace rgw::sal;
+
+	int ret =
+	  s->object->list_parts(
+            this, s->cct, max_parts, marker,
+	    &next_marker, &truncated,
+	    [&](const Object::Part& part) -> int {
+	      s->formatter->open_object_section("Part");
+	      s->formatter->dump_int("PartNumber", part.part_number);
+	      s->formatter->dump_unsigned("Size", part.part_size);
+	      if (part.cksum.type != rgw::cksum::Type::none) {
+		s->formatter->dump_string(part.cksum.element_name(), part.cksum.to_armor());
+	      }
+	      s->formatter->close_section(); /* Part */
+	      return 0;
+	    }, s->yield);
+
+	if (ret < 0) {
+	  ldpp_dout_fmt(this, 0,
+			"ERROR: {} list-parts failed for {}",
+			__func__, s->object->get_name());
+	}
+	/* AWS docs disagree on the name of this element */
+	s->formatter->dump_int("PartsCount", *multipart_parts_count);
+	s->formatter->dump_int("TotalPartsCount", *multipart_parts_count);
+	s->formatter->dump_bool("IsTruncated", truncated);
+	s->formatter->dump_int("MaxParts", max_parts);
+	s->formatter->dump_int("NextPartNumberMarker", next_marker);
+	s->formatter->dump_int("PartNumberMarker", marker);
+	s->formatter->close_section();
+      } /* multipart_parts_count positive */
+    } /* ObjectParts */
+
+    if (requested_attributes & as_flag(ReqAttributes::ObjectSize)) {
+      s->formatter->dump_int("ObjectSize", s->obj_size);
```

Comment:
i think we're relying on our call to `RGWGetObj::execute()` to set the correct size here, but you'd added an early-return for RGW_OP_GET_OBJ_ATTRS before we check the compression attrs for the real size: https://github.com/ceph/ceph/blob/main/src/rgw/rgw_op.cc#L2437-L2439

i commented out that early return and added a dout statement to log the value of `s->obj_size` in that `need_decompress` block, and it's printing the correct size:
> s3:get_obj_attrs setting real size=67108864 of compressed object

but the test still fails with `assert 347528 == 67108864`. something later in the request must be overwriting it?

Reply:
interesting, well it's more important to get the attributes than return early

Reply:
as it is now, we are only summing the lead object, if there exist more than one (as in this case?)

Reply:
(or is that ok?)

Reply:
> but the test still fails with assert 347528 == 67108864. something later in the request must be overwriting it?

sorry, i was looking at the wrong thing! this _does_ fix the test:
```diff
    /* STAT ops don't need data, and do no i/o */
-  if ((get_type() == RGW_OP_STAT_OBJ) ||
-      (get_type() == RGW_OP_GET_OBJ_ATTRS)) {
+  if (get_type() == RGW_OP_STAT_OBJ) {
     return;
```

Reply:
> interesting, well it's more important to get the attributes than return early

that early return was also skipping our handling of server-side encryption, which we want

> as it is now, we are only summing the lead object, if there exist more than one (as in this case?)

we're calling `get_part_obj_state()` to read the attrs from each part head object. the sizes in `RGW_ATTR_MANIFEST` and `RGW_ATTR_COMPRESSION` do account for part tail objects. is that what you were asking?

Reply:
> > interesting, well it's more important to get the attributes than return early
> 
> that early return was also skipping our handling of server-side encryption, which we want

dropping the early return wasn't enough to support sse-c, but i did get it working in https://github.com/linuxbox2/ceph/compare/wip-object-attributes...cbodley:ceph:pr-55259